### PR TITLE
Client statistics needs to be communicated to the cluster

### PR DESCRIPTION
**Pull request by ihsandemir**
Implemented the client statistics. Introduces two configuration parameters: hazelcast.client.statistics.enabled and hazelcast.client.statistics.period.seconds.

### CODE REVIEW

File: hazelcast-client/src/main/java/com/hazelcast/client/impl/HazelcastClientInstanceImpl.java [hunk 1]
```diff
@@ -242,6 +244,8 @@ public HazelcastClientInstanceImpl(ClientConfig config,
         lockReferenceIdGenerator = new ClientLockReferenceIdGenerator();
         nearCacheManager = clientExtension.createNearCacheManager();
         clientExceptionFactory = initClientExceptionFactory();
+
+        statistics = new Statistics(this);
```
Comment (pveentjer) [inline hazelcast-client/src/main/java/com/hazelcast/client/impl/HazelcastClientInstanceImpl.java]: Be careful with passing this in a constructor to another object. Easy to run into a partial constructed object (one of the Effective Java anti patterns)

File: hazelcast-client/src/main/java/com/hazelcast/client/impl/statistics/Statistics.java [hunk 2]
```diff
@@ -0,0 +1,235 @@
...
+
+        staticStats.put("/enterprise", Boolean.toString(enterprise));
+        // Register for one time statistics. These are resent only when reconnected to the cluster.
+        registerStaticStatistics();
+
+        registerPeriodicMetrics();
+
+        logger.finest("Client statistics is enabled with period " + periodSeconds + " seconds.");
+
+        // Start task for sending periodic statistics
```
Comment (pveentjer) [inline hazelcast-client/src/main/java/com/hazelcast/client/impl/statistics/Statistics.java]: If you need a comment for the line below, perhaps it makes more sense to just rename the method to the comment :)

startTaskForSendingPeriodicStatistics.

File: hazelcast-client/src/main/java/com/hazelcast/client/impl/statistics/Statistics.java [hunk 3]
```diff
@@ -0,0 +1,235 @@
...
+        scheduleStatisticsCollection(periodSeconds);
+    }
+
+    private void scheduleStatisticsCollection(long periodSeconds) {
+        client.getExecutionService().scheduleWithRepetition(new Runnable() {
+            @Override
+            public void run() {
+                final Map<String, String> stats = new HashMap<String, String>();
+
+                metricsRegistry.render(new ProbeRenderer() {
```
Comment (pveentjer) [inline hazelcast-client/src/main/java/com/hazelcast/client/impl/statistics/Statistics.java]: I think this approach is going to lead to problems because you have no control on the amount of probes that are going to be rendered. There could be a huge number; e.g. a client connecting to many servers; then for each connection there will easily be like 10/20 probes.

As we discussed, I believe it is best if you access the concrete probes you are looking for. Retrieve them once and recycle them, like is done in the HealthMonitor.HealthMetrics

File: hazelcast-client/src/main/java/com/hazelcast/client/impl/statistics/Statistics.java [hunk 4]
```diff
@@ -0,0 +1,235 @@
...
+                        logger.finest("Exception while processing statistics " + name + ". Error:" + e.getMessage());
+                    }
+
+                    @Override
+                    public void renderNoValue(String name) {
+                    }
+                });
+
+                // put user executor queue size
+                ThreadPoolExecutor userExecutor = (ThreadPoolExecutor) client.getClientExecutionService().getUserExecutor();
```
Comment (pveentjer) [inline hazelcast-client/src/main/java/com/hazelcast/client/impl/statistics/Statistics.java]: Probably better to get this apparently important value defined as a probe. Now you only get this data in case of client statistics it is not available in regular diagnostics.

File: hazelcast/src/main/java/com/hazelcast/client/ClientEndpoint.java [hunk 5]
```diff
@@ -105,6 +107,17 @@ void authenticated(ClientPrincipal principal, Credentials credentials, boolean f
      *
      * @param version The version string as obtained from the environment
      */
-    void setClientVersion(String version);
+    void setClientVersion(java.lang.String version);
```
Comment (pveentjer) [inline hazelcast/src/main/java/com/hazelcast/client/ClientEndpoint.java]: Are you expecting to import any other String? :) I would get rid of the java.lang

File: hazelcast-client/src/main/java/com/hazelcast/client/impl/statistics/Statistics.java [hunk 6]
```diff
@@ -0,0 +1,235 @@
...
+import com.hazelcast.spi.properties.HazelcastProperty;
+
+import java.util.HashMap;
+import java.util.Map;
+import java.util.concurrent.ThreadPoolExecutor;
+
+import static java.util.concurrent.TimeUnit.SECONDS;
+
+/**
+ * Created by ihsan on 15/05/17.
```
Comment (pveentjer) [inline hazelcast-client/src/main/java/com/hazelcast/client/impl/statistics/Statistics.java]: We can find the author using the git history.

File: hazelcast-client/src/main/java/com/hazelcast/client/impl/statistics/Statistics.java [hunk 7]
```diff
@@ -0,0 +1,235 @@
...
+ */
+public class Statistics {
+    private final boolean enabled;
+    private final HazelcastProperties properties;
+    private final ILogger logger = Logger.getLogger(this.getClass());
+    private final HazelcastClientInstanceImpl client;
+    private final boolean enterprise;
+    public static final String PREFIX = "hazelcast.client.statistics";
```
Comment (pveentjer) [inline hazelcast-client/src/main/java/com/hazelcast/client/impl/statistics/Statistics.java]: Checkstyle is going to choke on this. Statics should be defined first.

File: hazelcast-client/src/main/java/com/hazelcast/client/impl/statistics/Statistics.java [hunk 8]
```diff
@@ -0,0 +1,235 @@
...
+
+    public final static String NEAR_CACHE_CATEGORY_PREFIX = "/nearcache";
+
+    private final MetricsRegistry metricsRegistry = new MetricsRegistryImpl(Logger.getLogger(this.getClass()),
```
Comment (pveentjer) [inline hazelcast-client/src/main/java/com/hazelcast/client/impl/statistics/Statistics.java]: This seems a bit strange to me. Why does statistics create its own MetricsRegistry? Why don't you make use of the one that is available in the HazelcastClientInstanceImpl

File: hazelcast-client/src/main/java/com/hazelcast/client/impl/statistics/Statistics.java [hunk 9]
```diff
@@ -0,0 +1,235 @@
...
+            ProbeLevel.MANDATORY);
+    Map<String, String> staticStats = new HashMap<String, String>(1);
```
Comment (pveentjer) [inline hazelcast-client/src/main/java/com/hazelcast/client/impl/statistics/Statistics.java]: No access modifier even not final?

File: hazelcast-client/src/main/java/com/hazelcast/client/impl/statistics/Statistics.java [hunk 10]
```diff
@@ -0,0 +1,235 @@
...
+     * The period in seconds the statictics runs.
+     * <p>
+     * The StatisticsPlugin periodically sending the client statistics to the owner member
+     * <p>
+     * <p>
+     * If set to 0, the plugin is disabled.
+     */
+    public static final HazelcastProperty PERIOD_SECONDS = new HazelcastProperty(PREFIX + ".period.seconds", 60, SECONDS);
+
+    public Statistics(HazelcastClientInstanceImpl client) {
```
Comment (pveentjer) [inline hazelcast-client/src/main/java/com/hazelcast/client/impl/statistics/Statistics.java]: I would not inject HazelcastClientInstanceImpl, but the concrete dependencies to prevent coupling with the HazelcastClientInstanceImpl.

File: hazelcast-client/src/main/java/com/hazelcast/client/impl/statistics/Statistics.java [hunk 11]
```diff
@@ -0,0 +1,235 @@
...
+
+                // addNearCacheStats
+                addNearCachStats(stats);
+
+                sendStats(stats);
+            }
+        }, 0, periodSeconds, SECONDS);
+    }
+
+    private void addNearCachStats(Map<String, String> stats) {
```
Comment (pveentjer) [inline hazelcast-client/src/main/java/com/hazelcast/client/impl/statistics/Statistics.java]: Rendering to a StringBuffer would safe a lot of intermediate string and support objects from being created.

File: hazelcast-client/src/main/java/com/hazelcast/client/impl/statistics/Statistics.java [hunk 12]
```diff
@@ -0,0 +1,235 @@
...
+    }
+
+    private void registerStaticStatistics() {
+        // The client may have already connected, hence, send the stat at least once at startup
+        sendStats(staticStats);
+
+        client.getLifecycleService().addLifecycleListener(new LifecycleListener() {
+            @Override
+            public void stateChanged(LifecycleEvent event) {
+                if (LifecycleEvent.LifecycleState.CLIENT_CONNECTED.equals(event.getState())) {
```
Comment (pveentjer) [inline hazelcast-client/src/main/java/com/hazelcast/client/impl/statistics/Statistics.java]: I normally do static import of the enum to reduce horizontal noise.

File: hazelcast-client/src/main/java/com/hazelcast/client/impl/statistics/Statistics.java [hunk 13]
```diff
@@ -0,0 +1,235 @@
...
+
+    private void sendStats(Map<String, String> oneTimeStats) {
+        ClientMessage request = ClientStatisticsCodec.encodeRequest(oneTimeStats.entrySet());
+        Address ownerConnectionAddress = client.getClientClusterService().getOwnerConnectionAddress();
+        if (null != ownerConnectionAddress) {
+            try {
+                new ClientInvocation(client, request, ownerConnectionAddress).invoke().get();
+            } catch (Exception e) {
+                // suppress exception, do not print too many messages
+                logger.finest("Could not send stats " + e.getMessage());
```
Comment (pveentjer) [inline hazelcast-client/src/main/java/com/hazelcast/client/impl/statistics/Statistics.java]: Best to add a logging guard and to add the exception as cause. This way you can at least see the exception if somebody puts logging on finest. Discarding stacktraces makes it a pita to look for a problem

File: hazelcast-client/src/main/java/com/hazelcast/client/impl/statistics/Statistics.java [hunk 14]
```diff
@@ -0,0 +1,235 @@
...
+            }
+        }
+    }
+
+    private void registerPeriodicMetrics() {
+        RuntimeMetricSet.register(metricsRegistry);
+        OperatingSystemMetricSet.register(metricsRegistry);
```
Comment (pveentjer) [inline hazelcast-client/src/main/java/com/hazelcast/client/impl/statistics/Statistics.java]: This should not be needed if would not create your own MetricsRegistry

File: hazelcast-client/src/main/java/com/hazelcast/client/impl/statistics/Statistics.java [hunk 15]
```diff
@@ -0,0 +1,235 @@
...
+
+        // register user executor queue size
+        metricsRegistry
```
Comment (pveentjer) [inline hazelcast-client/src/main/java/com/hazelcast/client/impl/statistics/Statistics.java]: This need probe you define, you should define it on the client version so we get this new probe in general diagnostics

File: hazelcast-client/src/main/java/com/hazelcast/client/impl/statistics/Statistics.java [hunk 16]
```diff
@@ -0,0 +1,235 @@
...
+                .register("User Executor", "/userExecutor/queueSize", ProbeLevel.MANDATORY, new LongProbeFunction<String>() {
```
Comment (pveentjer) [inline hazelcast-client/src/main/java/com/hazelcast/client/impl/statistics/Statistics.java]: Why did you decide to use '/' as delimiter and not a  '.'

Now we have another style in the system.

File: hazelcast-client/src/test/java/com/hazelcast/client/statistics/ClientStatisticsTest.java [hunk 17]
```diff
@@ -0,0 +1,248 @@
...
+        Assert.assertNull(cacheHits);
+
+        IMap<Integer, Integer> map = client.getMap(testMapName);
+
+        // Wait enough time for statistics collection
+        sleepSeconds(statsPeriodSeconds + 1);
+
+        stats = getStats(client, clientEngine);
+        mapHits = stats.get("/nearcache/" + testMapName + "/Hits");
+        Assert.assertNotNull(mapHits);
```
Comment (pveentjer) [inline hazelcast-client/src/test/java/com/hazelcast/client/statistics/ClientStatisticsTest.java]: use static imports for asserts

File: hazelcast-client/src/main/java/com/hazelcast/client/impl/HazelcastClientInstanceImpl.java (hunk 1 above)
Comment (ihsandemir) [inline hazelcast-client/src/main/java/com/hazelcast/client/impl/HazelcastClientInstanceImpl.java]: I initially had the explicit parameters passed but it came out to be five of them (ClientExecutionService, ClusterService, LifeCycleService, NearCacheManager, ClientProperties) which I thought would be cleaner to pass "this" instead of five parameters. Partial construction will not be a problem for this case but I understand the concern. In addition, "this" is passed HazelcastClientInstanceImpl as a parameter to a number of objects. Do you think it is still better to pass five parameters instead of "this"?

File: hazelcast-client/src/main/java/com/hazelcast/client/impl/statistics/Statistics.java (hunk 2 above)
Comment (ihsandemir) [inline hazelcast-client/src/main/java/com/hazelcast/client/impl/statistics/Statistics.java]: OK, changed the name to schedulePeriodicStatisticsSendTask.

File: hazelcast-client/src/main/java/com/hazelcast/client/impl/statistics/Statistics.java (hunk 4 above)
Comment (ihsandemir) [inline hazelcast-client/src/main/java/com/hazelcast/client/impl/statistics/Statistics.java]: Yes, I agree, we can add this to Diagnostics. It will be better to send a separate PR for this change, since changing the Diagnostics related statistics is not in the scope of this PR.

File: hazelcast/src/main/java/com/hazelcast/client/ClientEndpoint.java (hunk 5 above)
Comment (ihsandemir) [inline hazelcast/src/main/java/com/hazelcast/client/ClientEndpoint.java]: Ok, corrected. I think IDE refactoring caused it and I missed this one.

File: hazelcast-client/src/main/java/com/hazelcast/client/impl/statistics/Statistics.java (hunk 6 above)
Comment (ihsandemir) [inline hazelcast-client/src/main/java/com/hazelcast/client/impl/statistics/Statistics.java]: OK, Removed.

File: hazelcast-client/src/main/java/com/hazelcast/client/impl/statistics/Statistics.java (hunk 8 above)
Comment (ihsandemir) [inline hazelcast-client/src/main/java/com/hazelcast/client/impl/statistics/Statistics.java]: The set of statistics to be collected and sent for client statistics is different from the diagnostics set of metrics, hence, i kept them as separate.

File: hazelcast-client/src/main/java/com/hazelcast/client/impl/statistics/Statistics.java (hunk 7 above)
Comment (ihsandemir) [inline hazelcast-client/src/main/java/com/hazelcast/client/impl/statistics/Statistics.java]: OK, corrected.

File: hazelcast-client/src/main/java/com/hazelcast/client/impl/statistics/Statistics.java (hunk 9 above)
Comment (ihsandemir) [inline hazelcast-client/src/main/java/com/hazelcast/client/impl/statistics/Statistics.java]: OK, made private final.

File: hazelcast-client/src/main/java/com/hazelcast/client/impl/statistics/Statistics.java (hunk 11 above)
Comment (ihsandemir) [inline hazelcast-client/src/main/java/com/hazelcast/client/impl/statistics/Statistics.java]: Ok, changed to use StringBuffer.

File: hazelcast-client/src/main/java/com/hazelcast/client/impl/statistics/Statistics.java (hunk 12 above)
Comment (ihsandemir) [inline hazelcast-client/src/main/java/com/hazelcast/client/impl/statistics/Statistics.java]: Ok, changed.

File: hazelcast-client/src/main/java/com/hazelcast/client/impl/statistics/Statistics.java (hunk 13 above)
Comment (ihsandemir) [inline hazelcast-client/src/main/java/com/hazelcast/client/impl/statistics/Statistics.java]: OK, changed.

File: hazelcast-client/src/main/java/com/hazelcast/client/impl/statistics/Statistics.java (hunk 14 above)
Comment (ihsandemir) [inline hazelcast-client/src/main/java/com/hazelcast/client/impl/statistics/Statistics.java]: Need our own metrics set.

File: hazelcast-client/src/main/java/com/hazelcast/client/impl/statistics/Statistics.java (hunk 15 above)
Comment (ihsandemir) [inline hazelcast-client/src/main/java/com/hazelcast/client/impl/statistics/Statistics.java]: Will do in another PR to add to diagnostics.

File: hazelcast-client/src/main/java/com/hazelcast/client/impl/statistics/Statistics.java (hunk 16 above)
Comment (ihsandemir) [inline hazelcast-client/src/main/java/com/hazelcast/client/impl/statistics/Statistics.java]: OK, changing it to be the same a Diagnostics usage.

File: hazelcast-client/src/main/java/com/hazelcast/client/impl/statistics/Statistics.java (hunk 11 above)
Comment (Donnerbart) [inline hazelcast-client/src/main/java/com/hazelcast/client/impl/statistics/Statistics.java]: If you don't need thread-safety, please use `StringBuilder` instead. `StringBuffer` is the same, but with synchronized methods.

File: hazelcast-client/src/test/java/com/hazelcast/client/statistics/ClientStatisticsTest.java (hunk 17 above)
Comment (ihsandemir) [inline hazelcast-client/src/test/java/com/hazelcast/client/statistics/ClientStatisticsTest.java]: OK, changed.

File: hazelcast-client/src/main/java/com/hazelcast/client/impl/statistics/Statistics.java (hunk 3 above)
Comment (ihsandemir) [inline hazelcast-client/src/main/java/com/hazelcast/client/impl/statistics/Statistics.java]: OK, changed.

File: hazelcast-client/src/main/java/com/hazelcast/client/impl/HazelcastClientInstanceImpl.java (hunk 1 above)
Comment (pveentjer) [inline hazelcast-client/src/main/java/com/hazelcast/client/impl/HazelcastClientInstanceImpl.java]: Passing concrete dependencies is in almost all cases a better approach.

File: hazelcast-client/src/main/java/com/hazelcast/client/impl/statistics/Statistics.java (hunk 11 above)
Comment (pveentjer) [inline hazelcast-client/src/main/java/com/hazelcast/client/impl/statistics/Statistics.java]: Jit probably gets rid of the locks and even if it couldn't it would be very cheap because there is only a single thread needing the 'lock'. But yes.. StringBuilder is better.. somehow I keep forgetting it.

File: hazelcast-client/src/main/java/com/hazelcast/client/impl/statistics/Statistics.java (hunk 14 above)
Comment (pveentjer) [inline hazelcast-client/src/main/java/com/hazelcast/client/impl/statistics/Statistics.java]: Having a single shared set of metrics and just filtering out the elements you need make more sense. Duplication a pain. 

Also I would not grab all probes; but just the focus on the most important ones.

File: hazelcast-client/src/main/java/com/hazelcast/client/impl/HazelcastClientInstanceImpl.java (hunk 1 above)
Comment (ihsandemir) [inline hazelcast-client/src/main/java/com/hazelcast/client/impl/HazelcastClientInstanceImpl.java]: @pveentjer I tried not using "this" but I end up needing it due to this line: https://github.com/ihsandemir/hazelcast/blob/clientStats/hazelcast-client/src/main/java/com/hazelcast/client/impl/statistics/Statistics.java#L196  ClientInvocation constructor needs ClientInstance, hence it seems like I can not get rid of it unless ClientInvocation is also changed. We can open a cleanup issue for this since this change affects a number of different places ("this" is used a lot in HazelcastClientInstanceImpl).

File: hazelcast-client/src/main/java/com/hazelcast/client/impl/statistics/Statistics.java (hunk 11 above)
Comment (ihsandemir) [inline hazelcast-client/src/main/java/com/hazelcast/client/impl/statistics/Statistics.java]: OK, changed.

File: hazelcast-client/src/main/java/com/hazelcast/client/impl/statistics/Statistics.java (hunk 14 above)
Comment (ihsandemir) [inline hazelcast-client/src/main/java/com/hazelcast/client/impl/statistics/Statistics.java]: Adding filtering capability to the MetricsRegistry would be a nice feature but I think we need to do it in another PR.

File: hazelcast/src/main/java/com/hazelcast/client/impl/ClientEngineImpl.java [hunk 18]
```diff
@@ -588,6 +590,32 @@ public Operation getPostJoinOperation() {
         return resultMap;
     }
 
+    @Override
+    public List<Map.Entry<String, List<Map.Entry<String, String>>>> getClientStatistics() {
+        Collection<ClientEndpoint> clientEndpoints = endpointManager.getEndpoints();
+        List<Map.Entry<String, List<Map.Entry<String, String>>>> statsList = new ArrayList<Map.Entry<String, List<Map.Entry<String, String>>>>(
+                clientEndpoints.size());
+        for (final ClientEndpoint e : clientEndpoints) {
+            statsList.add(new Map.Entry<String, List<Map.Entry<String, String>>>() {
```
Comment (sancar) [inline hazelcast/src/main/java/com/hazelcast/client/impl/ClientEngineImpl.java]: Instead of Entry interface use AbstractMap.SimpleEntry it also has correct hashcode and equals implementations which is missing here.

Comment (ihsandemir) [inline hazelcast/src/main/java/com/hazelcast/client/impl/ClientEngineImpl.java]: OK, done.

File: hazelcast-client/src/main/java/com/hazelcast/client/impl/HazelcastClientInstanceImpl.java (hunk 1 above)
Comment (pveentjer) [inline hazelcast-client/src/main/java/com/hazelcast/client/impl/HazelcastClientInstanceImpl.java]: I know the problem. In some parts on the serverside we also have this issue where 'this' is being passed to constructor.

File: hazelcast-client/src/main/java/com/hazelcast/client/impl/statistics/Statistics.java [hunk 19]
```diff
@@ -0,0 +1,246 @@
...
+    public final static int FEATURE_SUPPORTED_SINCE_VERSION = BuildInfo.calculateVersion(FEATURE_SUPPORTED_SINCE_VERSION_STRING);
+
+    private final boolean enabled;
+    private final HazelcastProperties properties;
+    private final ILogger logger = Logger.getLogger(this.getClass());
+    private final HazelcastClientInstanceImpl client;
+
+    private final boolean enterprise;
+
+    private final MetricsRegistry metricsRegistry = new MetricsRegistryImpl(Logger.getLogger(this.getClass()),
```
Comment (pveentjer) [inline hazelcast-client/src/main/java/com/hazelcast/client/impl/statistics/Statistics.java]: I'm not convinced this is the right approach. Creating a new instance just for statistics smells like a design problem. Also keep in mind that a MetricsRegistry has its own thread; so by creating another instance you add another thread to the system.

So I would reuse the existing metrics and filter out all the probes that are mandatory; should be part of the PRD. Minimize this so we can add new ones in the future.

File: hazelcast-client/src/main/java/com/hazelcast/client/impl/statistics/Statistics.java [hunk 20]
```diff
@@ -0,0 +1,246 @@
...
+        if (null == ownerConnectionAddress) {
+            return null;
+        }
+        ClientConnection connection = (ClientConnection) client.getConnectionManager().getConnection(ownerConnectionAddress);
+        if (null == connection) {
+            return null;
+        }
+
+        int serverVersion = connection.getConnectedServerVersion();
+        if (serverVersion < FEATURE_SUPPORTED_SINCE_VERSION) {
```
Comment (pveentjer) [inline hazelcast-client/src/main/java/com/hazelcast/client/impl/statistics/Statistics.java]: Could it be that this is going to be logged potentially every x seconds if someone has enabled finest?

File: hazelcast-client/src/main/java/com/hazelcast/client/impl/statistics/Statistics.java [hunk 21]
```diff
@@ -0,0 +1,246 @@
...
+                final Map<String, String> stats = new HashMap<String, String>();
+
+                stats.put("lastStatisticsCollectionTime", Long.toString(System.currentTimeMillis()));
+
+                stats.putAll(staticStats);
+
+                periodicStats.fillMetrics(stats);
+
+                ThreadPoolExecutor userExecutor = (ThreadPoolExecutor) client.getClientExecutionService().getUserExecutor();
+                stats.put("userExecutor.queueSize", Long.toString(userExecutor.getQueue().size()));
```
Comment (pveentjer) [inline hazelcast-client/src/main/java/com/hazelcast/client/impl/statistics/Statistics.java]: This should have been exposed as a probe so that also the regular diagnostics can read out this value. THis way the design doesn't get riddles with ad hoc fixes.

File: hazelcast-client/src/main/java/com/hazelcast/client/impl/statistics/Statistics.java [hunk 22]
```diff
@@ -0,0 +1,246 @@
...
+            if (name.startsWith("/")) {
+                pathPrefix += name.substring(1);
+            } else {
+                pathPrefix += name;
+            }
+            pathPrefix += ".";
+
+            NearCacheStats nearCacheStats = nearCache.getNearCacheStats();
+
+            StringBuilder buffer = new StringBuilder();
```
Comment (pveentjer) [inline hazelcast-client/src/main/java/com/hazelcast/client/impl/statistics/Statistics.java]: Why is the buffer created in the loop? If you create it outside the loop, you get less litter.

Comment (ihsandemir) [inline hazelcast-client/src/main/java/com/hazelcast/client/impl/statistics/Statistics.java]: OK, changed.

File: hazelcast/src/main/java/com/hazelcast/internal/metrics/metricsets/RuntimeMetricSet.java [hunk 23]
```diff
@@ -97,5 +106,10 @@ public long get(RuntimeMXBean runtimeMXBean) {
                     }
                 }
         );
+        registeredMetrics.add("runtime.uptime");
+    }
+
+    public static Set<String> getRegisteredMetricNames() {
```
Comment (pveentjer) [inline hazelcast/src/main/java/com/hazelcast/internal/metrics/metricsets/RuntimeMetricSet.java]: I would just check out the probes as in HealthMonitor.HealthMetrics.

File: hazelcast-client/src/main/java/com/hazelcast/client/impl/statistics/Statistics.java (hunk 20 above)
Comment (ihsandemir) [inline hazelcast-client/src/main/java/com/hazelcast/client/impl/statistics/Statistics.java]: yes, that is correct. Alternatives:
1. No logging: But then how can we know if this was the reason of not getting the statistics populated.
2. Do the log only once per connection: But then this would require unnecessary lookup table and such.

Do you have any other suggestion?

1.

File: hazelcast-client/src/main/java/com/hazelcast/client/impl/statistics/Statistics.java (hunk 21 above)
Comment (ihsandemir) [inline hazelcast-client/src/main/java/com/hazelcast/client/impl/statistics/Statistics.java]: Which metric set do we want to put it to?

File: hazelcast/src/main/java/com/hazelcast/internal/metrics/metricsets/RuntimeMetricSet.java (hunk 23 above)
Comment (ihsandemir) [inline hazelcast/src/main/java/com/hazelcast/internal/metrics/metricsets/RuntimeMetricSet.java]: In HealthMetrics, the metric names are being duplicated. Is this what you like, duplicating the names? All I need is to get all the statistics that are registered for RuntimeMetricSet and OperatingSystemMetricSet.

File: hazelcast-client/src/main/java/com/hazelcast/client/impl/statistics/Statistics.java (hunk 21 above)
Comment (pveentjer) [inline hazelcast-client/src/main/java/com/hazelcast/client/impl/statistics/Statistics.java]: You can define a probe directly on the metrics system.

```
 <S> void register(S source, String name, ProbeLevel level, LongProbeFunction<S> probe);
```

File: hazelcast/src/main/java/com/hazelcast/internal/metrics/metricsets/RuntimeMetricSet.java (hunk 23 above)
Comment (pveentjer) [inline hazelcast/src/main/java/com/hazelcast/internal/metrics/metricsets/RuntimeMetricSet.java]: This duplication is fine because these are the keys of which the probes can be used; it is their official contract.

Comment (pveentjer) [inline hazelcast/src/main/java/com/hazelcast/internal/metrics/metricsets/RuntimeMetricSet.java]: If you really want to be lazy; just walk over all probes and filter out the probes you don't want based on their name. If it doesn't start with e.g. 'os' then you ignore it. The probes you like you store in a hashmap.

File: hazelcast-client/src/main/java/com/hazelcast/client/impl/statistics/Statistics.java (hunk 21 above)
Comment (ihsandemir) [inline hazelcast-client/src/main/java/com/hazelcast/client/impl/statistics/Statistics.java]: OK, done.

File: hazelcast/src/main/java/com/hazelcast/internal/metrics/metricsets/RuntimeMetricSet.java (hunk 23 above)
Comment (ihsandemir) [inline hazelcast/src/main/java/com/hazelcast/internal/metrics/metricsets/RuntimeMetricSet.java]: OK, changed to use hard coded statistic names.

File: hazelcast/src/main/java/com/hazelcast/internal/metrics/StringGauge.java [hunk 24]
```diff
@@ -0,0 +1,36 @@
+/*
+ * Copyright (c) 2008-2017, Hazelcast, Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
```
Comment (pveentjer) [inline hazelcast/src/main/java/com/hazelcast/internal/metrics/StringGauge.java]: Why is a new gauge needed? I don't think this is the right approach.

File: hazelcast-client/src/main/java/com/hazelcast/client/impl/statistics/Statistics.java [hunk 25]
```diff
@@ -0,0 +1,254 @@
...
+            if (name.startsWith("/")) {
+                pathPrefix += name.substring(1);
+            } else {
+                pathPrefix += name;
+            }
+            pathPrefix += ".";
+
+            NearCacheStats nearCacheStats = nearCache.getNearCacheStats();
+
+            stats.put(buffer.append(pathPrefix).append("creationTime").toString(),
```
Comment (pveentjer) [inline hazelcast-client/src/main/java/com/hazelcast/client/impl/statistics/Statistics.java]: I see a lot of litter being created in this code. Lots of Strings and underlying byte-arrays.

File: hazelcast-client/src/main/java/com/hazelcast/client/impl/statistics/Statistics.java [hunk 26]
```diff
@@ -0,0 +1,254 @@
...
+
+        private final Map<String, StringGauge> allMetrics = new HashMap<String, StringGauge>(STATISTIC_NAMES.length);
+
+        PeriodicStatistics(final MetricsRegistry metricsRegistry) {
+            for (String name : STATISTIC_NAMES) {
+                allMetrics.put(name, metricsRegistry.newStringGauge(name));
+            }
+        }
+
+        void fillMetrics(final Map<String, String> metricsMap) {
```
Comment (pveentjer) [inline hazelcast-client/src/main/java/com/hazelcast/client/impl/statistics/Statistics.java]: I'm questioning the type of the 'metricsMap'.

If you would just send a formatted String to the other side, then only a single String is needed.

File: hazelcast-client/src/main/java/com/hazelcast/client/spi/impl/ClientExecutionServiceImpl.java [hunk 27]
```diff
@@ -103,6 +108,11 @@ public ExecutorService getUserExecutor() {
         return userExecutor;
     }
 
+    @Probe (level = MANDATORY)
```
Comment (pveentjer) [inline hazelcast-client/src/main/java/com/hazelcast/client/spi/impl/ClientExecutionServiceImpl.java]: Very nice. 

You are allowed to return int; queue.size will also return int.

File: hazelcast-client/src/main/java/com/hazelcast/client/impl/statistics/Statistics.java (hunk 20 above)
Comment (pveentjer) [inline hazelcast-client/src/main/java/com/hazelcast/client/impl/statistics/Statistics.java]: Another approach would be to limit the rate of this exception to be printed. E.g. print it at most once a minute.

File: hazelcast/src/main/java/com/hazelcast/internal/metrics/StringGauge.java (hunk 24 above)
Comment (ihsandemir) [inline hazelcast/src/main/java/com/hazelcast/internal/metrics/StringGauge.java]: There is no object creation after start for the gauge. It is initialized and reused, hence it should not cause any litter. The gauge is cached in Statistics.PeriodicStatistics object. 

The reason why defined the string gauge is this: in construction at PeriodicStatistics I only know the name of the gauge but do not know if the type is long or double, hence StringGauge parses the value to string correctly based on the gauge type.

File: hazelcast-client/src/main/java/com/hazelcast/client/spi/impl/ClientExecutionServiceImpl.java (hunk 27 above)
Comment (ihsandemir) [inline hazelcast-client/src/main/java/com/hazelcast/client/spi/impl/ClientExecutionServiceImpl.java]: OK, changed.

File: hazelcast-client/src/main/java/com/hazelcast/client/impl/statistics/Statistics.java (hunk 20 above)
Comment (ihsandemir) [inline hazelcast-client/src/main/java/com/hazelcast/client/impl/statistics/Statistics.java]: OK, cached the last connected server address and only printing the message if the last cached Address is different from the current connected server.

File: hazelcast-client/src/main/java/com/hazelcast/client/connection/nio/ClientConnection.java [hunk 28]
```diff
@@ -79,6 +80,7 @@ public ClientConnection(HazelcastClientInstanceImpl client,
         channel.attributeMap().put(ClientConnection.class, this);
         this.connectionId = connectionId;
         this.logger = client.getLoggingService().getLogger(ClientConnection.class);
+        this.startTime = System.currentTimeMillis();
```
Comment (pveentjer) [inline hazelcast-client/src/main/java/com/hazelcast/client/connection/nio/ClientConnection.java]: Flavor warning: 

I try to initialize fields when I can to reduce the size of the constructor.

Comment (ihsandemir) [inline hazelcast-client/src/main/java/com/hazelcast/client/connection/nio/ClientConnection.java]: OK, updated.

File: hazelcast/src/main/java/com/hazelcast/internal/metrics/StringGauge.java (hunk 24 above)
Comment (ihsandemir) [inline hazelcast/src/main/java/com/hazelcast/internal/metrics/StringGauge.java]: OK, removed StringGauge and implemented render method in the Long and Double gauges as suggested.

**Comment by ihsandemir** [comment]
run-lab-run
